### ISSUE TITLE
Question: correctly handle deleted topics with regex consumer

### ISSUE THREAD

**Issue by alistairking** [twmb/franz-go#434]
I have a regex consumer and it seems like something isn't quite working as I'd expect when I delete some topics out from under it. There's a few symptoms, but even hours after the deletion, I'm still seeing `assigning partitions` logs that include the deleted partitions. I'm also seeing lots of error logs from the Kafka brokers like
```
org.apache.kafka.common.errors.UnknownTopicOrPartitionException: Error while fetching partition state for my_partition
```
(although weirdly these logs didn't start showing up until I added some _new_, differently named, topics).

I came across `PurgeTopicsFromClient`, but I don't (think I) have any way for the client to know which topics should be purged when this happens.

What am I missing here?

**Comment by twmb** [comment]
If a topic is missing from a metadata response, it's unclear (per the Kafka protocol) if the topic was deleted or if the broker being queried is stale. Because of this, the client doesn't trust a topic missing from the response indicating something about deletion.

For regex consuming, rather than requesting specific topics, the client always requests all topics and then filters them. So, even disregarding the above, there's no indication in a metadata response that a topic is missing (via the UNKNOWN_TOPIC_OR_PARTITION error).

So, with regex consuming, once a topic is discovered as consumable, the client permanently attempts to consume it, and there currently is no means within the client to automatically discover that a topic was deleted.

I could probably go through some effort to internally purge the topic if _every_ broker indicates the topic no longer exists. This does make more sense to do in the face of regex consuming. If my prior effort of adding `ConsumeRecreatedTopics` is any indication (which I reverted because there are just too many edge cases), it might not be trivial. If you can use `PurgeTopicsFromClient`, that's more robust since it punts responsibility to the user (you) that a topic is actually deleted. Do you have a means of broadcasting to clients that a topic was deleted?

It is surprising to me that you're only seeing those errors after a new, differently named topic -- the client should always be trying to fetch topics it is assigned / currently knows of.

**Comment by alistairking** [comment]
Ah, very interesting. Do you have any idea how rdkafka handles this? (I only ask because I know this "worked" -- or at least appeared to -- when I was using rdkafka.)

I'll have a think about how I could broadcast delete information, but in my setup deletes are pretty infrequent, so it might be easier to just do a rolling restart of clients after a delete operation.

Alternatively... Is there a way we could have the clients opt into being less careful and treating a topic that goes missing as deleted? What's the corner case that would result in the metadata from a broker missing a topic that actually exists (other than shortly after the topic is created?)

**Comment by twmb** [comment]
It looks like librdkafka immediately trusts a topic to be deleted if it's first (a) known to the client and then (b) disappears from a metadata response. So, there's logic within to handle this case (vs. the franz-go current behavior of not trusting brokers).

**Comment by twmb** [comment]
I'm going to try to handle internally purging deleted topics, but no ETA yet. I hope to set aside time for it this weekend.

**Comment by alistairking** [comment]
That'd be great. Thanks. Let me know if there's anything I can do to help.

**Comment by twmb** [comment]
This has been fixed in v1.13.3, please let me know if it does not work for you.

**Comment by alistairking** [comment]
Amazing. Thank you!
